### PR TITLE
bazel: Add `json_data` and writers

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: bazel: Add json_data and providers
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: bazel/api_binding.bzl [hunk 1]
```diff
@@ -13,6 +13,7 @@ def _default_envoy_api_impl(ctx):
     ]
     for d in api_dirs:
         ctx.symlink(ctx.path(ctx.attr.envoy_root).dirname.get_child(ctx.attr.reldir).get_child(d), d)
+    ctx.symlink(ctx.path(ctx.attr.envoy_root).dirname.get_child("bazel").get_child("utils.bzl"), "utils.bzl")
```
Comment (phlax) [inline bazel/api_binding.bzl:16]: ive tried a few different ways to make the `json_data` macro available in the api part of the repo

not sure the best way

Comment (htuch) [reply bazel/api_binding.bzl:16]: Possibly the cleanest way is to put the `json_data` macro inside `api/bazel`. That makes the dependency chain simpler.

**Review by htuch — DISMISSED** [review]
Nice!

File: bazel/api_binding.bzl (hunk 1 above)
Comment (phlax) [reply bazel/api_binding.bzl:16]: not sure why i didnt think of that - ill move it

Comment (phlax) [reply bazel/api_binding.bzl:16]: ...in a follow up...

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**